dvc/remote/gdrive/pydrive.py

```diff
@@ -0,0 +1,104 @@
+import os
+
+from dvc.remote.gdrive.utils import TrackFileReadProgress, FOLDER_MIME_TYPE
+
+
+class RequestBASE:
```

Comment:
Hello @Suor ,
There is some decisions history leaded to creation of `RequestBASE` classes:
- `PyDrive` is used to simplify interaction with Google API and to avoid our custom implementation as it was done in previous PR ( https://github.com/iterative/dvc/pull/2040/files )
- It is turned out that `PyDrive` doesn't provide API requests refreshing mechanism( in case of 403, 50X error codes. Have asked PyDrive's devs for help there - https://github.com/gsuitedevs/PyDrive/issues/177 .  ) There is Google's API recommendations what kind of errors should be refreshed https://developers.google.com/drive/api/v2/manage-uploads#errors ( 403 and 50X errors ).
- To implement request refreshing mechanism for PyDrive `RequestBASE` classes were introduced. Each `RequestBASE` class encapsulate single call of underlying Google API accessed through PyDrive internals. Exact call happens inside overloaded `def execute(self):`.

> These are functions masquerading as classes complicating code.

Could you please give more details how exactly the code is complicated with these classes?
Initial implementation was done without classes and introduction of classes slightly decreased the complexity of source code (by my opinion): 1. Specific RequestBASE instance is created and passed to `RemoteGDrive.execute_request`. 2. PyDrive's internals have less impact on organization of source code inside RemoteGDrive.

I agree that RequestBASE classes don't have internal state (except `RequestListFilePaginated`) and `functions` should suite better, but current implementation helps to minimize "monolithic" implementation of `RemoteGDrive`. If introduction of new classes is not acceptable at all, I would still try to move PyDrive's specific calls outside of RemoteGDrive. Btw, in case of functions, functions will start accept pretty long list of input arguments (which currently are passed to objects on init).

Please double confirm that these classes should be killed, thanks

Reply:
@Suor This is extension to pydrive library. It's good to have it as a separate file. Good to have aligned interfaces (not sure if it's the case here). Would love to hear more arguments besides just "no-no". Don't consider this as major blocker. Let's focus on fixing the external credentials interface first.

Reply:
So,  `RequestListFile` and `RequestCreateFolder` might be simply inlined:
- copy their `.execute()` contents to their use place
- slap a `@gdrive_retry` decorator on the enclosing method (see below)

For `RequestUploadFile` and `RequestDownloadFile` there is a question whether we can safely cache path ids for all path not only root. Is that special somehow? If the answer is yes, then:
- cache `.get_path_id()` call
- inline these same as above

If no then it is the same as with `RequestListFilePaginated`.

For `RequestListFilePaginated` we will need to isolate the api call, somewhat like:
```python
from funcy.py3 import cat

def list_drive_item(self, query):
    file_list = self.drive.ListFile({"q": query, "maxResults": 1000})
    
    # Isolate and decorate the op we want to retry
    get_list = gdrive_retry(lambda: next(file_list, None))
    
    # Fetch pages until None is received, lazily flatten the thing
    return cat(iter(get_list, None))
```

The `@gdrive_retry` should do 2 things:
- intercept pydrive exceptions we want to retry and wrap them
- retry

So it's a composition:
```python
from funcy import compose, decorator, retry

@decorator
def _wrap_pydrive_retriable(call):
    try:
        return call()
    except SomePyDriveError as e:
        if <should be retried>:
            raise GDriveRemoteRetriableError(..., cause=e)
        else:
            raise

gdrive_retry = compose(
    retry(8, GDriveRemoteRetriableError, timeout=...),
    _wrap_pydrive_retriable
)
```

Reply:
Some of this code might look dense, but it will:
- keep the core implementation of each method local
- keep retry and wrap behavior separate and somewhat declarative
- no need to pass through, assign and fetch all the args
- less code)

Reply:
the 1st check I've checked is if you are @Suor the author of `funcy` and bingo! :D 
will try to apply your suggestions :+1:

Reply:
@Suor @MaxRis can it be considered resolved?